userspace/libsinsp/metrics_collector.h

```diff
@@ -0,0 +1,317 @@
+	 * exposed at /metrics in Falco's existing HTTP webserver (w/ optional mTLS support), eliminating the need for implementing 
+	 * a complete Prometheus client.
+	 * 
+	 * We exclusively support counter and gauge Prometheus metric types, covering metrics from kernel driver tracepoints
+	 * to linsinsp and client metrics. Introducing a registry seems excessive, especially given the dynamic nature of the final
+	 * metric string names, such as variations in tracepoints across architectures.
+	 * Considering the simplistic use case, adding another dependency to the project does not seem justified. Furthermore, for C++ 
+	 * (compared to Go for example), there appear to be fewer formal client library projects available. Plus, we need to think 
+	 * about stability and long-term support before adding any new dependency.
+	 * 
+	 * The final fully qualified Prometheus metric name partially follows https://prometheus.io/docs/practices/naming/
+	 * Prepend namespace and subsystem with "_" delimiter to create a fully qualified metric name according to 
+	 * https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Opts + append unit with "_" delimiter
+	 * We do not strictly follow and enforce the concept of base_units, but guarantee no units are mixed per unique 
+	 * `prometheus_metric_name_fully_qualified`
+	 * 
+	 * We are monitoring updates wrt https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md
+	 * 
+	 * Example:
+	 * 
+	 * # HELP testns_falco_n_threads_total https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_n_threads_total gauge
+	 * testns_falco_n_threads_total{raw_name="n_threads",example_key1="example1",example_key2="example2"} 12
+	 * # HELP testns_falco_memory_rss_megabytes https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_memory_rss_megabytes gauge
+	 * testns_falco_memory_rss_megabytes{raw_name="memory_rss",example_key1="example1",example_key2="example2"} 350.000000
+	 * 
+	 * This method is a work in progress.
+	 * 
+	 * @param metric metrics_v2 metric
+	 * @param prometheus_namespace first component of `prometheus_metric_name_fully_qualified` (optional)
+	 * @param prometheus_subsystem second component of `prometheus_metric_name_fully_qualified` (optional)
+	 * @param const_labels map of additional labels (rarely used for a metrics_v2 metric)
+	 * @return Complete new line delimited text-based Prometheus exposition format metric string 
+	 * w/ a `prometheus_metric_name_fully_qualified` - optional components prepended to and unit appended to. 
+	 * 3-lines including # HELP and # TYPE lines followed by the metric line, raw metric name always present as label.
+	*/
+	std::string convert_metric_to_text(metrics_v2 metric, std::string_view prometheus_namespace = "", std::string_view prometheus_subsystem = "", const std::map<std::string,std::string>& const_labels = {}) const;
```

Comment:
We initially discussed that we should have a universal `convert_metric_to_text` for the different subclasses. Prometheus however uses additional special params compared to what you would use to define the base. Now sure renaming it will alleviate it, but it would defy the initial design purpose. I noticed this being suboptimal during dev ...

What would you suggest we do?

When I chatted with @sgaist earlier today he was guessing maybe do something in line of

```
std::string convert_metric_to_text(const metrics_v2& metric) override { return convert_metric_to_prometheus(metric); }
std::string convert_metric_to_prometheus(const metrics_v2& metric, std::string_view prometheus_namespace = "", std::string_view prometheus_subsystem = "", const std::map<std::string,std::string>& const_labels = {}) const;
```

Awaiting additional clarifications from your side @federico-sysdig. This would be crucial to get sorted out.

Reply:
I don't understand. A virtual function is usually a way to invoke an action on an object without knowing exactly its specific type. If you have a (non-virtual) function that is an overload of a virtual function you can't call it using a base class pointer to the object, you have to know the specific type. In this case it should not be a problem calling a function with a different name.

```c++
struct Base {
    virtual void foo() {}
};

struct Derived1 {
    void foo() override {}
};

struct Derived2 {
    void foo() override {}
    void foo(int) {} // overload, not override!
};

int main(int argc, char* argv{}) {
    Base* ptr = argv[1][0] == '2' ? new Derived2 : new Derived1;

    ptr->foo(); // ok, the actual function to call is resolved at runtime
    ptr->foo(5); // try calling the overload, not the virtual, but it's not possible on Base
}
```

Reply:
A different name it is then! Thanks!